server/sonar-process/src/main/java/org/sonar/process/ProcessEntryPoint.java

```diff
@@ -100,26 +100,46 @@ public void launch(Monitored mp) {
       stopWatcher.start();
 
       monitored.start();
-      boolean up = false;
-      while (!up) {
-        up = monitored.isUp();
-        Thread.sleep(20L);
-      }
-
-      // notify monitor that process is ready
-      commands.setUp();
-
-      if (lifecycle.tryToMoveTo(Lifecycle.State.STARTED)) {
-        monitored.awaitStop();
+      Monitored.Status status = waitForNotDownStatus();
+      if (status == Monitored.Status.UP || status == Monitored.Status.OPERATIONAL) {
+        // notify monitor that process is ready
+        commands.setUp();
+
+        if (lifecycle.tryToMoveTo(Lifecycle.State.STARTED)) {
+          waitForOperational(status);
+          if (lifecycle.tryToMoveTo(Lifecycle.State.OPERATIONAL)) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS134)